### ISSUE TITLE
HTML for Godot Beta 10 Takes 1-2 Minutes to Load on MacOS

### ISSUE THREAD

**Issue by Luzzotica** [godotengine/godot#70691]
### Godot version

4.0.beta10

### System information

MacOS Monterey 12.2.1

### Issue description

Godot games block the main thread (And potentially crashes it) on MacOS.

Untested for Windows or Linux.

This happens from both a generic build and when you run the project in a browser from the engine.

### Steps to reproduce

Create a new project for Godot Beta 10.

Add a UI Button to the screen.

Have it print things out, whatever you wish.

Add an HTML5 export, install the HTML5 export library if necessary.

Click the run on HTML5 button on the top right.

Proceed to wait for 1-2 minutes before the game will fully load.

### Minimal reproduction project

[TestGodot4Web.zip](https://github.com/godotengine/godot/files/10317128/TestGodot4Web.zip)

**Comment by Calinou** [comment]
Duplicate of https://github.com/godotengine/godot/issues/41118 and/or https://github.com/godotengine/godot/issues/68647 (same cause – large size causes slow initialization times).

**Comment by Luzzotica** [comment]
@Calinou I feel like this isn't the same as those. I don't believe it's a size issue.

Running locally on MacOS on both Brave and Chrome: Took about ~2 minutes and 30 seconds for the game to load. Both Brave and Chrome wanted to kill the tab. It had locked up the entire browser for (Couldn't go to any other tab).

Running this on Windows, and the game loads in ~10 seconds. 

If this is a size issue, then MacOS Chrome and Brave are just INSANELY slow to load things. Which doesn't seem right. 
Also, since I'm running it locally, I assume the web assembly can load faster... This might be untrue, but I doubt that it is.

**Comment by Zireael07** [comment]
In that case, it might be shader loading and/or conversion (correct me if I'm wrong, but MacOS converts stuff to ANGLE behind the scenes, doesn't it?)

**Comment by clayjohn** [comment]
> In that case, it might be shader loading and/or conversion (correct me if I'm wrong, but MacOS converts stuff to ANGLE behind the scenes, doesn't it?)

I also think may be a shader compilation issue. Chrome definitely uses ANGLE, but I am not sure about Brave. macOS doesn't use ANGLE when running on desktop though. 

While attempting to fix the gl_compatibility renderer on macOS I experimented with using ANGLE, but found that the shader compilation times ended up increasing by over 100X. In https://github.com/godotengine/godot/pull/70065 I was able to speed up shader compilation times and get gl_compatibility working on macOS without ANGLE. However, since chrome uses ANGLE, my guess is we are hitting those pathological shader compilation issues anyway. I am not sure what exactly in the shader is causing the issue (it may even be a bug in ANGLE).

**Comment by bergice** [comment]
Related https://github.com/godotengine/godot/issues/65696

**Comment by thomasbergersen** [comment]
The issue of Chrome crashing on MacOS 13 does indeed exist, and Firefox takes over 2 minutes to load.

**Comment by shinspiegel** [comment]
I'm completely new to C++ dev, but I would like to help with this issue. 
What can I do to help?

**Comment by Luzzotica** [comment]
Can confirm, 4.0 RC2 is still taking minutes to load on Brave on MacOS.. Causes my entire browser to lock down completely.

**Comment by dhnm** [comment]
The issue still persists in rc4, rc5, rc6. In Forward+, Mobile, and Compability mode. Chrome hangs, Firefox takes a while but loads, and Safari reports the following error in the console:

```[Error] Unhandled Promise Rejection: CompileError: WebAssembly.Module doesn't parse at byte 1155: can't get 1th argument Type
	promiseEmptyOnRejected
	promiseReactionJob
> Selected Element
< <html lang="en">…</html>
[Error] still waiting on run dependencies:
	onPrintError (tmp_js_export.js:14008)
	(anonymous function) (tmp_js_export.js:737)
[Error] dependency: wasm-instantiate
	onPrintError (tmp_js_export.js:14008)
	(anonymous function) (tmp_js_export.js:739)
[Error] (end of list)
	onPrintError (tmp_js_export.js:14008)
	(anonymous function) (tmp_js_export.js:742)```

**Comment by michaldev** [comment]
On latest godot 4 (rc 6) it still blocks web browsers (tested on chrome and firefox) on macos. Will there really be a stable version of godot released today with this bug?

**Comment by max-vogler** [comment]
Just to confirm - this issue does occur on Godot 4.0 stable, viewing the web export in Chrome 110 on macOS 13.2 on a MacBook Pro 2023 M2. A project containing only a very simple 2D pixel art scene takes 70s to load initially. Is it possible to find a workaround until this is fixed, e.g. strip shaders from the web export?

**Comment by Calinou** [comment]
> Is it possible to find a workaround until this is fixed, e.g. strip shaders from the web export?

No, other than staying on Godot 3.x.

**Comment by greggman** [comment]
I extracted the shaders from godot 4.0 from the sample linked above and filed a bug. Only 19 shaders are compiled at startup but for some reason (didn't look into why) several shaders, 5 shaders take ~15 seconds each for a total of 1 minute at 15 seconds (on an M1 mac)

https://bugs.chromium.org/p/angleproject/issues/detail?id=8068

**Comment by greggman** [comment]
If you're curious I wrote this code

```js
// extract-shader.js
(function() {
  const programs = [];

  function dumpPrograms(label) {
    if (programs.length) {
      if (label) {
        console.log('//', label);
      }
      console.log(JSON.stringify(programs));
      programs.length = 0;
    }
  }

  globalThis.dumpPrograms = dumpPrograms;

  for (const RenderingContext of [WebGL2RenderingContext, WebGLRenderingContext]) {
    RenderingContext.prototype.linkProgram = function(origFn) {
      return function(prg) {
        console.log('// link', programs.length);
        const shaders = this.getAttachedShaders(prg);
        programs.push(shaders.map(sh => {
          return {
            type: this.getShaderParameter(sh, this.SHADER_TYPE),
            src: this.getShaderSource(sh),
          };
        }));
        origFn.call(this, prg);
      };
    }(RenderingContext.prototype.linkProgram);

    for (const name of ['clear', 'drawArray', 'drawElements', 'drawArraysInstanced', 'drawElementsInstanced']) {
      RenderingContext.prototype[name] = function(origFn) {
        return function(...args) {
          dumpPrograms();
          origFn.call(this, ...args);
        };
      }(RenderingContext.prototype[name]);
    }
  }
})();
```

And then I added `<script src="extract-shaders.js'></script>` to the top of the html file and I added `importScripts('extrac-shaders.js')` to the worker (as I wasn't sure which one uses WebGL, the main thread, the worker, or both). I also added a call to `dumpShaders` in the `onmessage` code in the worker and in the `onPrint` function in the main thread JavaScript since the sample above does so little that I wasn't sure WebGL is actually being called.

In any case, you can use code like this to extract shaders and then code like the one in the chromium bug report to make a minimal repo.

**Comment by greggman** [comment]
So looking into this more, it appears to be an issue with ANGLE running on top of the MacOS OpenGL drivers. Chrome is working to switch over to ANGLE running on top of Metal. You can test that path with

/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal --user-data-dir=<some-random-folder>

In which case the shaders compile. Hopefully that will ship soon.

You can verify which backend ANGLE is running on by going to `about:gpu`

ANGLE running on OpenGL:

> GL_RENDERER: ANGLE (Apple, Apple M1 Max, OpenGL 4.1 Metal - 83)

ANGLE running on Metal:

> GL_RENDERER: ANGLE (Apple, ANGLE Metal Renderer: Apple M1 Max, Version 13.2.1 (Build 22D68))

Yes, it's confusing that both strings say 'metal' 😅

**Comment by greggman** [comment]
> So we just gotta wait for chrome/safari/brave to migrate to the new Metal Renderer? 

correct

**Comment by evwilliams** [comment]
Is it possible to fallback to the way Godot 3.x handled web exports while we wait for browsers to migrate to Metal?

**Comment by Calinou** [comment]
> Is it possible to fallback to the way Godot 3.x handled web exports while we wait for browsers to migrate to Metal?

No, as 3.x featured entirely different rendering backends (most notably a GLES2 backend, which is much simpler and more limited).

**Comment by tavurth** [comment]
Probably a worse case of this https://github.com/godotengine/godot/issues/56670 (3.x)

**Comment by Xwaffle1** [comment]
Hi I just wanted to share my frustration with the bug reported on https://github.com/godotengine/godot/issues/72584. In 3.x exporting to Web was a safe way to ensure your game was playable on many platforms. 

Web exports for a very minimal 2D game are unable to be played in chrome on a MAC, however interestingly enough Safari seems to work. It's very unfortunate because web exports being cross platform was a huge selling point for me.

I look forward to progress on this.

**Comment by Calinou** [comment]
@Xwaffle1 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by pbantolas** [comment]
This looks fixed on the latest macOS Ventura (13.3.1) using Safari (16.4). I'm running this on a 2021 M1 Macbook Pro. I tested a variety of scenes both with Godot 4.0.2 and Godot 3.5.2 with the WebGL2 export. Safari uses Metal and not ANGLE for WebGL2 and perhaps that's why the behaviour is different.

The only error I get in the console is
`[Error] USER WARNING: Project setting: rendering/limits/global_shader_variables/buffer_size exceeds maximum uniform buffer size of: 16384`
which is mitigated by changing the project setting to the recommended value.

Firefox still hangs for a bit before loading the scene. Running a CPU Profile reveals that most of the work is stalling on some sort of shader introspection from a GL library. Which means that Firefox is running on a GL layer.

![SCR-20230421-ucoh](https://user-images.githubusercontent.com/821868/233743329-bed34c21-ea55-4047-9a31-cbdf92374fa6.png)

The recommended Chrome workaround (https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280) works for me.

In general there is quite a bit of variability, but given the majority of people are probably using Chrome this is still dependent on them switching to the Metal backend.

**Comment by AThousandShips** [comment]
Won't get any focus until after 4.1 is released, only high impact bugs and crashes get much focus at that stage, unless this is a trivial fix and someone finds it pretty much immediately

**Comment by AThousandShips** [comment]
If the issue is open it hasn't been fixed :) no PR has been linked to this so no fix seems to be in progress right now, keep looking here for updates like that

**Comment by mgc8** [comment]
> If the issue is open it hasn't been fixed :) no PR has been linked to this so no fix seems to be in progress right now, keep looking here for updates like that

I guess the confusion comes from the fact that issues such as [this](https://github.com/tgreaves/avenger-sharks/issues/10) leave a strong indication that it would be worked on for 4.1. While that is not clearly not the case, perhaps a better indication of a timeline for investigation/fix could be added to the documentation, just so people embarking on potential new projects can plan accordingly?

**Comment by AThousandShips** [comment]
How does an issue on a separate repository indicate anything about the priorities in this one? This issue is clearly marked "4.x", and no PR has been opened for this so no solution is clear.

> perhaps a better indication of a timeline for investigation/fix could be added to the documentation

We don't track these things in the documentation, that's what the issues here are for

**Comment by clayjohn** [comment]
> > If the issue is open it hasn't been fixed :) no PR has been linked to this so no fix seems to be in progress right now, keep looking here for updates like that
> 
> I guess the confusion comes from the fact that issues such as [this](https://github.com/tgreaves/avenger-sharks/issues/10) leave a strong indication that it would be worked on for 4.1. While that is now clearly not the case, perhaps a better indication of a timeline for investigation/fix could be added to the documentation, just so people embarking on potential new projects can plan accordingly?

It might be helpful for you to circle back on [previous comments](https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280) in this thread. The issue here is with Chrome's implementation of ANGLE on MacOS. We are waiting for them to resolve the issue upstream. The Godot Project has no influence over Google's release cycles, nor do we have special insight into them. So we unfortunately can't give you a timeline when Google will be shipping a fix, nor can we speed up the timeline.

**Comment by akien-mga** [comment]
I would suggest for someone affected by this issue to make a clear MRP and report the bug upstream to Chromium: https://bugs.chromium.org/p/chromium/issues/list

It would be important to include information from that comment: https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280
i.e. that `--use-angle=metal` solves the issue, but is sadly not the default behavior as of current Chrome releases on macOS.

They hopefully track this internally, but some user reports confirming that the current status quo is not satisfactory can help them prioritize. I see some existing upstream issues about slow WebGL performance on Chrome macOS, but not specific to Godot, so it could be good to add one.

Another option is the more "end user bug report" pipeline with https://support.google.com/chrome/answer/95315?hl=en&co=GENIE.Platform%3DDesktop - a bunch of users each reporting that a readily accessible online project runs terribly on Chrome while it runs well on Safari and Firefox will likely help prioritize further.

**Comment by mgc8** [comment]
> How does an issue on a separate repository indicate anything about the priorities in this one? This issue is clearly marked "4.x", and no PR has been opened for this so no solution is clear.

Apologies for the confusion, that was more on the path of arriving here from discussions on various platforms; it was all about setting the right expectations, since [posts on Twitter](https://twitter.com/reduzio/status/1623253494021988353) and the [documentation itself](https://docs.godotengine.org/en/stable/tutorials/export/exporting_for_web.html) leave the impression that this is either a temporary issue or something being worked on.

If that is not the case and it's all dependent on Google fixing Chrome and Mozilla fixing Firefox, then that's all fine and good to know, but it needs to me made clear so we go and fill the right bugs and follow-ups for those platforms as necessary (as well pointed out below by @akien-mga). For example, the documentation currently mentions "_Safari has several issues with WebGL 2.0 support that other browsers don't have, so we recommend using a Chromium-based browser or Firefox if possible_", which is exactly opposite at the moment, with Safari being the only functional browser in macOS. Perhaps a note there that Safari works and Chrome/Firefox need upstream fixes would help point people in the right direction?

Thank you and sorry again for any confusion.

**Comment by soundgnome** [comment]
> If that is not the case and it's all dependent on Google fixing Chrome and Mozilla fixing Firefox

It isn't though, because web builds made with Godot 3.5.2 don't have this problem. Maybe at some point changes to Firefox and Chrome will make the issue go away, but I rather doubt developers working on those projects consider addressing Godot regressions to be their job.

**Comment by akien-mga** [comment]
@soundgnome Try GLES3 on Godot 3, you should have the same issue. It's a WebGL 2 support issue in Chromium, which relies on deprecated OpenGL drivers for ANGLE instead of the modern Metal backend that Apple developed and contributed to Google for ANGLE.

**Comment by mgc8** [comment]
> I would suggest for someone affected by this issue to make a clear MRP (both a ready-to-test demo available online, and the source with instructions on how to export using Godot) and report the bug upstream to Chromium: https://bugs.chromium.org/p/chromium/issues/list

Thank you for the suggestion, there appear to be actually two issues there that reference this problem:
1. One opened by @greggman from the above comments in March -- https://bugs.chromium.org/p/angleproject/issues/detail?id=8068
2. An older one referencing issues with Godot and WASM specifically (but having to do with WebGL); rather than creating a new issue, I've added a test project and steps for reproducing to that second issue -- https://bugs.chromium.org/p/chromium/issues/detail?id=1324296

Hopefully, this will attract the attention of the right people who can either fix the problem or work on switching to the Metal renderer.

Thank you!

**Comment by Morpheu5** [comment]
I stirred up a wasps nest – sorry! But at least we have pointers to go poke Google in the right places so thank you! :)

**Comment by kevzettler** [comment]
I don't understand why this issue is being deferred to the Google and the browser vendors. Obviously this is a breaking change on the Godot project. If web exports work in 3.5.2 this is a breaking change. The Godot project scrapped the stable WebGL1.0 export pipeline in favor of broken WebGL2.0. Its fairly well known that WebGL2.0 is not yet stable or widely adopted with browser vendors. 

Ideally I would assume here that Godot would offer a stable WebGL1.0 exporter from the 3.5.2 line and then introduce a new unstable WebGL2.0 exporter. Is it possible to restore the WebGL1.0 pipeline? I'm assuming this not trivial because during the 3 -> 4 major version change some architectural changes.

**Comment by Calinou** [comment]
> Ideally I would assume here that Godot would offer a stable WebGL1.0 exporter from the 3.5.2 line and then introduce a new unstable WebGL2.0 exporter. Is it possible to restore the WebGL1.0 pipeline?

This would mean recreating a GLES2 renderer from scratch and maintaining it separately from the existing GLES3 and Vulkan renderers (no code can be reused). We don't have enough rendering contributors to be able to maintain that.

**Comment by SephReed** [comment]
> ```
> /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal --user-data-dir=<some-random-folder>
> ```

every <some-random-folder> I've tried hasn't worked.  Does somebody have a more specific example?

**Comment by rteijeiro** [comment]
> > ```
> > /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal --user-data-dir=<some-random-folder>
> > ```
> 
> every `<some-random-folder>` I've tried hasn't worked. Does somebody have a more specific example?
> 
> <img alt="image" width="261" src="https://user-images.githubusercontent.com/14002040/252126327-ce7d9d7b-a6e2-4339-a514-e7374d64b2cf.png">

That's a permissions issue in your Mac configuration. This works fine at least in my computer.

**Comment by hunky-dory** [comment]
Respectfully, this isn't simply an issue of the game taking a while to load — running a Godot 4 web game causes my entire browser to become unresponsive until I manage to kill the tab, which is hard to do when it causes simple tasks like changing tabs, opening right click menus or closing tabs to take 8-12 seconds.

If solving the issue is reliant on fixes over which you have no control, might it be possible to warn OSX Chrome users before starting the loading process?

**Comment by Calinou** [comment]
> If solving the issue is reliant on fixes over which you have no control, might it be possible to warn OSX Chrome users before starting the loading process?

Given [Chrome user agents are now frozen](https://groups.google.com/a/chromium.org/g/blink-dev/c/-2JIRNMWJ7s/m/yHe4tQNLCgAJ), there's no longer a way to warn users of a specific browser version without also impacting users with a newer browser version that has the bug fixed.

**Comment by nickarora** [comment]
For what it's worth, you can use javascript to identify whether the browser is running [Angle on OPEN GL](https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280)

```
const gl = document.createElement("canvas").getContext("webgl");
const debugInfo = gl.getExtension('WEBGL_debug_renderer_info');
gl.getParameter(debugInfo.UNMASKED_RENDERER_WEBGL);

// => 'ANGLE (Apple, Apple M1 Pro, OpenGL 4.1)'
```

It's not a great solution, but you could potentially prevent your application from loading if it detects an incompatibility.  I haven't tested this in practice.

**Comment by mgc8** [comment]
> For what it's worth, you can use javascript to identify whether the browser is running [Angle on OPEN GL](https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280)

Unfortunately, it's not enough to test for ANGLE, but for the specific case of using the Metal renderer -- a @greggman notes [above](https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280):

> ANGLE running on OpenGL:
> GL_RENDERER: ANGLE (Apple, Apple M1 Max, OpenGL 4.1 Metal - 83)
> ANGLE running on Metal:
> GL_RENDERER: ANGLE (Apple, ANGLE Metal Renderer: Apple M1 Max, Version 13.2.1 (Build 22D68))
> Yes, it's confusing that both strings say 'metal' 😅

But this might be useful to look for "ANGLE Metal Renderer" specifically, and perhaps show an alert or some red warning text? It's not about incompatibility, but slowness in loading (i.e. 1-2 minutes even, depending on shaders used), which causes the browser to "lock".

**Comment by fguillen** [comment]
So, for mac users, we can open Godot web games by opening Chrome with this command: 

    /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal

Note that you must make sure Chrome is completely 100% quit before running it this way or you can pass in another profile

    /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal --user-data-dir=<some-random-folder>

[Source](https://bugs.chromium.org/p/angleproject/issues/detail?id=8068#c2)

**Comment by mgc8** [comment]
> So, for mac users, we can open Godot web games by opening Chrome with this command:
(...)

Yes, this was taken from a comment [higher up](https://github.com/godotengine/godot/issues/70691#issuecomment-1456888280) here. However, in the meantime it has become less relevant, since developers on macOS can easily test using Safari -- which works fine and comes included with the OS.

The problem remains for end-users, who are not going to switch browsers or run arcane Terminal invocations in order to test a game or demo on e.g. itch.io, and that can only be solved by Chrome itself...

**Comment by azjgard** [comment]
> However, in the meantime it has become less relevant, since developers on macOS can easily test using Safari -- which works fine and comes included with the OS.

@mgc8 It seems like there's still a high degree of relevance to opening an alternative Chrome profile since opening in Safari suffers from the issue described [here](https://github.com/godotengine/godot/issues/74512) (which in turn links to [this comment](https://github.com/godotengine/godot/issues/70691#issuecomment-1446563071) from higher up in this thread). 

I'm not able to test the build of my simple game in normal Chrome or Safari; Chrome with the `--use-profile=metal` flag is the only way that I've been able to play an HTML5 export.

**Comment by mgc8** [comment]
> @mgc8 It seems like there's still a high degree of relevance to opening an alternative Chrome profile since opening in Safari suffers from the issue described [here](https://github.com/godotengine/godot/issues/74512) (which in turn links to [this comment](https://github.com/godotengine/godot/issues/70691#issuecomment-1446563071) from higher up in this thread).

The comment and bug referenced there are very old and pertain to macOS 13.2, while the current version is 13.4. Latest Safari (16.5) has no issues whatsoever playing the "MinimalProject.zip" included in report #74512:
<img width="1058" alt="Screenshot 2023-07-30 at 02 12 28" src="https://github.com/godotengine/godot/assets/7091920/b65d0812-11f5-4a4f-8844-5290f211a9fc">

> I'm not able to test the build of my simple game in normal Chrome or Safari; Chrome with the `--use-profile=metal` flag is the only way that I've been able to play an HTML5 export (alt: waiting several minutes for the FF build to load).

I can run my own projects as well as official demos without issues or delays in the current version of Safari. Do you have a build/demo of your game that fails we could test with?

As a side-note, new versions of Firefox (at least since around 112+) have also improved considerably in playing Godot games, the delay on load is only about 30 seconds now (with additional few seconds every time new shaders need to be compiled). The only big problem remains with Chrome, which indeed locks up and takes many minutes in the default (non-Metal) pathway.

**Comment by fguillen** [comment]
> As a side-note, new versions of Firefox (at least since around 112+) have also improved considerably

In my case:

- macOS: 13.1 (22C65) Intel Core i5
- Firefox: 115.0.3 (64-bit)

This project with just a few images takes 1.20seconds to load

- https://arturitoproductions.itch.io/testingimageissue (password: godot)

In Safari (16.2 (18614.3.7.1.5)), the same project above doesn't show any signal to load. I waited 2 minutes.

In iPad Chrome or Safari Godot4 web exported projects also don't load. I assume it is the same for iPhone

**Comment by mgc8** [comment]
> > As a side-note, new versions of Firefox (at least since around 112+) have also improved considerably
> In my case:
> * macOS: 13.1 (22C65) Intel Core i5
> * Firefox: 115.0.3 (64-bit)

Ouch, looks like you last updated that machine over six months ago! Current macOS Ventura version is 13.5 and Safari 16.5.

> This project with just a few images takes 1:20 minutes to load
> * https://arturitoproductions.itch.io/testingimageissue (password: godot)

It takes about 45 seconds for me -- but yes, Firefox is still slow and needs more fixes to be properly functional in this context. However, it used to completely lock the tab and/or error out, so the fact that it at least loads and works now is good progress.

> In Safari (16.2 (18614.3.7.1.5)), the same project above doesn't show any signal to load. I waited 2:00 minutes.

This is a known problem on itch.io specifically, because they insist on using the wrong headers for Safari, as reported here (assuming you've already enabled the SharedArrayBuffers support):
https://itch.io/t/2025776/experimental-sharedarraybuffer-support#post-7852760

<img width="459" alt="Screenshot 2023-07-31 at 01 00 47" src="https://github.com/godotengine/godot/assets/7091920/e7fcd088-ae68-4b67-af90-0d740c6662b9">

... but in your case that is a moot point, as you need to use a recent version of Safari anyway (the bugs discussed here were fixed in 16.4 which was released back in March):
https://bugs.webkit.org/show_bug.cgi?id=247860

You don't need to upload your games to itch.io in order to test functionality, you can do that locally much faster by running a simple HTTP server in the Web build directory (the one containing your "index.js" file) like the following, then using Safari to go to `http://localhost:8000`:
```
#!/usr/bin/env python3
from http import server # Python 3

class MyHTTPRequestHandler(server.SimpleHTTPRequestHandler):
        def end_headers(self):
                self.send_my_headers()
                server.SimpleHTTPRequestHandler.end_headers(self)

        def send_my_headers(self):
                self.send_header("Access-Control-Allow-Origin", "*")
                self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
                self.send_header("Cross-Origin-Opener-Policy", "same-origin")

if __name__ == '__main__':
        server.test(HandlerClass=MyHTTPRequestHandler)
```

> In iPad Chrome or Safari Godot4 web exported projects also don't load. I assume it is the same for iPhone

That is another kettle of fish altogether, i{pad}OS is not macOS and i{pad}OS Safari is not the same as the desktop version; plus there is no difference with Chrome in i{pad}OS, it's the just Safari but with a different UI. There currently is at least one known bug that prevents Godot 4 working properly with it -- the "Out Of Memory" one referenced in #70621 -- but the workaround presented in that issue can be used to make it functional: **manually setting the memory size to 256MiB instead of the default 2048MiB**. The secondary WebAssembly SIMD issue mentioned there was solved since 16.2 as well. You can use the same kind of script as above to test with that too, but there is an added complication that it needs SSL as well if not connecting to localhost (a relatively simple change).

> A bigger project like this (with shaders):
> * https://arturitoproductions.itch.io/coco-and-frogy
> In the above Firefox took ... don't know, I waited 3:00 minutes still loading

That's a fun game, by the way, refreshing take on the Arkanoid genre! It loaded in Firefox for me after about 2:12 minutes, with another delay caused by screen transitions (probably due to new shaders being loaded there each time). It plays well, but crashes after the tutorial with a console error (same in Chrome on Windows).

Here it is running perfectly in Safari though, with no crashes even (using the above localhost solution):
<img width="808" alt="Screenshot 2023-07-31 at 01 27 46" src="https://github.com/godotengine/godot/assets/7091920/1b89d887-529a-4ebe-a5b2-0153e7658ed6">

... and it also worked great on an iPad, by the way, using the memory workaround described above!

So, at this point in time at least, for macOS users and especially developers, Safari is the best bet for running Godot 4 games. itch.io needs to fix their headers, and Chrome needs to fix their default render pathways, for sure, but those are not really things Godot can influence much...

**Comment by Calinou** [comment]
> itch.io needs to fix their headers

Would https://github.com/godotengine/godot-proposals/issues/6616 be usable on itch.io as a workaround?

**Comment by mgc8** [comment]
> * https://arturitoproductions.itch.io/coco-and-frogy

It _should_, and from testing it does appear to set the right headers, but for some reason when loaded through itch.io the app still doesn't "see" them. Could be due to being inside an iFrame, or a conflict between the headers set by itch.io and the ones set in JavaScript?

As a side-note, while experimenting with this I noticed that if you right-click and open _just the iFrame for the game_ in a separate tab/window, the headers are set fine (with "require-corp") and it works nicely in Safari -- case in point, the game posted a bit earlier:
https://html.itch.zone/html/8415314/coco_and_frogy_web_v1.0.15/index.html

**Comment by mgc8** [comment]
> It _should_, and from testing it does appear to set the right headers, but for some reason when loaded through itch.io the app still doesn't "see" them. Could be due to being inside an iFrame, or a conflict between the headers set by itch.io and the ones set in JavaScript?

After some thorough testing, I think I tracked down the issues with `itch.io` (there were two):
1. `coi-serviceworker.js` itself has a bug that reverses the '_credentialless_' and '_request-corp_' options, thus (sigh) only affecting Safari, since Chrome and Firefox work with either option; reported [here](https://github.com/gzuidhof/coi-serviceworker/issues/20)
2. itch.io complicates the header situation due to the way the "game" pages are configured:
- If the "experimental SharedArrayBuffer" option is **not** selected, the main page will have no headers, while the iframe can set them fine (such as with our service-worker script), but the browser will not activate the "crossOriginIsolated" status because the parent container needs to have it as well. If the main page had been loaded on the same domain as the iframe, this would not have been an issue, since the service-worker would have taken that over as well (tested locally) and overridden the headers. But on itch.io they are on different hostnames, thus it can't touch the parent.
- If the experimental option is selected, then the main page will have headers, but hardcoded to "credentialless"; meanwhile, the iframe can override its own headers with the service-worker and send the correct ones for each browser (i.e. 'credentiallless' for Chrome and 'request-corp' for Safari). This will work fine in Chrome or Firefox, but Safari breaks because it doesn't see the parent page as being properly isolated due to the 'credentialless' there, which our service-worker again cannot override; thus it breaks.

So, for now, the "open iframe in its own tab/window" option remains as a workaround, while itch.io could improve the situation by adding a bit of browser-detection code to switch between 'credentialless' and 'request-corp' as needed, though I'm not sure if they have any inclination to do that.

**Comment by fguillen** [comment]
Thanks @mgc8, all this is great support for all the _suffering_ mac Godot users like me

**Comment by rosskarchner** [comment]
I've noticed that as of recently (I think) you can now set the backend using chrome://flags

![Screenshot 2023-08-28 at 2 48 36 PM](https://github.com/godotengine/godot/assets/235397/769c72af-447f-4d4f-8231-d6d6a2a66b14)


> So, for mac users, we can open Godot web games by opening Chrome with this command:
> 
> ```
> /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal
> ```
> 
> Note that you must make sure Chrome is completely 100% quit before running it this way or you can pass in another profile
> 
> ```
> /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --use-cmd-decoder=passthrough --use-angle=metal --user-data-dir=<some-random-folder>
> ```
> 
> [Source](https://bugs.chromium.org/p/angleproject/issues/detail?id=8068#c2)

**Comment by mgc8** [comment]
I'm not sure why it disappeared, but I can confirm that the Chrome flags suggestion posted earlier works great in Chrome 116.0.5845.179/macOS/M1, and is an easier temporary workaround rather than the Terminal version of the same (at least until the Chrome devs make Metal the default):
<img width="1200" alt="Screenshot 2023-09-09 at 21 50 47" src="https://github.com/godotengine/godot/assets/7091920/cf583512-1652-40de-866c-26cd131499df">

**Comment by leafo** [comment]
> So, for now, the "open iframe in its own tab/window" option remains as a workaround

On itchio, you should assume that at some point your game will be required to be loaded in an iframe that has been generated by itch.io. A solution that involves breaking out the frame is not a long term fix because we have system in place to automatically inject site locking to prevent shady third-party platforms from embedding our games to steal bandwidth. Additionally, we may expose game-to-site platform APIs over the `postMessage`.

> while itch.io could improve the situation by adding a bit of browser-detection code to switch between 'credentialless' and 'request-corp' as needed

Unfortunately `require-corp` has a bunch of side effects with regards to any external references on the page, and may impact how the site operates, so the solution isn't simply change the value of the header. This is why we opted for `credentialless` since it's mostly transparent solution, but the browser compatibility is not ideal.

I think our long term fix (assuming safari never supports the header) will probably involve launching the game in a separate page containing only an iframe nad different headers specific to safari, instead of trying to execute the game on the full project page.

**Comment by transmutrix** [comment]
On Newgrounds, SharedArrayBuffer support works in Safari as well as the other major browsers (tested with the same build of the same project, etc.). I'm not sure which method they've used, however, or if it would translate easily to itch, if it would be useful to look into, or who over there to ask about it if it *would* be useful.

**Comment by mgc8** [comment]
> Edit2: I deployed a fix that will cause Safari to always load games in a new window using the embed URL I described above that uses the `require-corp` header value. Should resolve any issues for Safari users, at the expense of slightly different user experience.

Hi @leafo, thank you for dropping by! This looks like a good, pragmatic solution given the current state of browsers, with every major one seemingly having a different, incompatible issue related to this. I can confirm that the "new window" works nicely for itch.io games in Safari, so at least that's solved now for end-users. Hopefully the browser developers will also move on with updates/fixes to make the whole process less frustrating for both site owners and game devs (three months later, Chrome still hasn't switched to Metal as default back-end on macOS)...

**Comment by transmutrix** [comment]
The "new window" solution works great on desktop! On mobile, it gives me a 404 error (tried iOS and iPadOS, both Safari and Chrome, but did not try Android).

**Comment by kamko0** [comment]
As of right now me using IOS doesn't seem to load my game that used Godot4, so even with the Embed game link it doesn't work.   
Not sure if there will be a fix anytime soon but at least it works on pc for now.

**Comment by revanj** [comment]
Hi there @leafo , thank you so much for the fix and for creating the website indie devs know and love!

In recent exports, the new window in safari appears to give 404. This seems to indicate that the generation of the embed link is not working correctly. Is there anything users can do to specify the embed link, or otherwise make the new window use the correct embed link?

**Comment by leafo** [comment]
> Hi there @leafo , thank you so much for the fix and for creating the website indie devs know and love!
> 
> In recent exports, the new window in safari appears to give 404. This seems to indicate that the generation of the embed link is not working correctly. Is there anything users can do to specify the embed link, or otherwise make the new window use the correct embed link? This is on a laptop M1 Mac.

Is your safari configured to block cookies in any way? A cookie is combined with the generated link to verify access to the popup

**Comment by leafo** [comment]
@revanj 

> In recent exports, the new window in safari appears to give 404. This seems to indicate that the generation of the embed link is not working correctly. Is there anything users can do to specify the embed link, or otherwise make the new window use the correct embed link? This is on a laptop M1 Mac.

Is your safari configured to block cookies in any way? A cookie is combined with the generated link to verify access to the popup

**Comment by tavurth** [comment]
> @revanj
> 
> > In recent exports, the new window in safari appears to give 404. This seems to indicate that the generation of the embed link is not working correctly. Is there anything users can do to specify the embed link, or otherwise make the new window use the correct embed link? This is on a laptop M1 Mac.
> 
> Is your safari configured to block cookies in any way? A cookie is combined with the generated link to verify access to the popup

My safari is fresh out of the box and shows the same issue.

<img width="265" alt="Screenshot 2023-11-28 at 22 21 01" src="https://github.com/godotengine/godot/assets/100964/18948669-9280-4764-82b6-28cbbac68bdb">

**Comment by adamscott** [comment]
It seems to be way better when exporting using the single-threaded export template. https://github.com/godotengine/godot/pull/85939

**Comment by Morpheu5** [comment]
Maybe on Chrome, but now it takes forever on Firefox/macOS.

**Comment by HamdyElzanqali** [comment]
Still the same issue, even with the `nothreads` template.

**Comment by nickarora** [comment]
As an update, it does appear that Chrome is now using the Metal renderer by default now.  I am using Chome 123.0.6312.59 and I have verified that the Angle Metal renderer is being used.  I have not changed any flags or set a custom profile.  

Also the "image rendering issue" renders without issue.
https://arturitoproductions.itch.io/testingimageissue (password: godot)

I also use Brave but it still not updated to use the Angle Metal Renderer.  I suspect its only a matter of time for the other browsers to follow suit.

**Comment by clayjohn** [comment]
> As an update, it does appear that Chrome is now using the Metal renderer by default now. I am using Chome 123.0.6312.59 and I have verified that the Angle Metal renderer is being used. I have not changed any flags or set a custom profile.
> 
> Also the "image rendering issue" renders without issue. https://arturitoproductions.itch.io/testingimageissue (password: godot)
> 
> I also use Brave but it still not updated to use the Angle Metal Renderer. I suspect its only a matter of time for the other browsers to follow suit.

Thank you very much for updating us! I suspect firefox and brave should update soon to match, but I worry that safari is still a long ways off. At any rate, having everything work well on the others puts us in a much better position.

**Comment by tavurth** [comment]
> Also the "image rendering issue" renders without issue. https://arturitoproductions.itch.io/testingimageissue (password: godot)

That demo locks up my Chrome, but no longer locks up my computer.

It will sit there like this for a long time:
<img width="740" alt="Screenshot 2024-03-27 at 10 03 43" src="https://github.com/godotengine/godot/assets/100964/8706630c-06ef-4fa0-8e8e-0584d99a5175">

<img width="274" alt="Screenshot 2024-03-27 at 10 05 04" src="https://github.com/godotengine/godot/assets/100964/d7bcc933-6ec6-4f04-bfce-bc5b69f9d435">

<img width="1207" alt="Screenshot 2024-03-27 at 10 06 10" src="https://github.com/godotengine/godot/assets/100964/a0404b15-7896-440d-9187-403e464f31a7">

Interesting that this appears to be a CPU bound issue, why is that?

**Comment by nickarora** [comment]
@tavurth The screenshot you posted indicates you need to relaunch Chrome for chrome version you've listed to take effect. I might try relaunching chrome and trying again.  The other difference is that I am using an ARM (M1) machine, but I can't see why that would matter.

Once you've restarted Chrome, can you open the console and run the following commands and share the output:
```
const gl = document.createElement("canvas").getContext("webgl");
const debugInfo = gl.getExtension('WEBGL_debug_renderer_info');
gl.getParameter(debugInfo.UNMASKED_RENDERER_WEBGL);
```
FWIW, on my machine, I see: 

```
ANGLE (Apple, ANGLE Metal Renderer: Apple M1 Pro, Unspecified Version)
```

**Comment by chrissearle** [comment]
That's interesting. Can I ask what chrome version?

Reason for asking - I have what chrome says is latest as of today: 123.0.6312.87 on an M3 macbook pro and I get:

`ANGLE (ATI Technologies Inc., AMD Radeon Pro 5300M OpenGL Engine, OpenGL 4.1)`

**Comment by starry-abyss** [comment]
On my M1 laptop the build with a few number labels from here https://arturitoproductions.itch.io/testingimageissue opens quickly on Chrome, but slowly on Firefox. And on Safari it's popping up a new itch.io window, which gives a 404 error :)

**Comment by nickarora** [comment]
Firefox has likely not adopted the Angle Metal renderer as their default yet.  

It may be the same with the latest version of non-ARM Chrome installs.  If the x86 (intel based) installs of chrome are not yet defaulted to Angle Metal renderer, then we'll have to wait a bit longer for Godot 4 web exports to work as expected

**Comment by nickarora** [comment]
It appears the new chrome default has only been applied to ARM based macs. intel based macs may have to wait a bit longer. Still, it's promising that these changes are slowly trickling out

**Comment by starry-abyss** [comment]
I believe this is fixed when using the new Godot 4.3 no-threads export.
